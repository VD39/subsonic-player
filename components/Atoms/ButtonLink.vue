<script setup lang="ts">
const props = withDefaults(
  defineProps<{
    disabled?: boolean;
    fullWidth?: boolean;
    icon?: Component | Icon;
    iconColor?: string;
    iconPosition?: IconPosition;
    iconSize?: IconSize;
    iconWeight?: IconWeight;
    is?: ButtonLink;
    showText?: boolean;
    type?: string;
  }>(),
  {
    disabled: undefined,
    icon: undefined,
    iconColor: undefined,
    iconPosition: 'left',
    iconSize: 'medium',
    iconWeight: 'regular',
    is: 'button',
    type: 'button',
  },
);

const type = props.is === 'button' ? props.type : undefined;
const isComponent =
  props.is === 'nuxt-link' ? resolveComponent('NuxtLink') : props.is;
</script>

<template>
  <component
    :is="isComponent"
    :class="[
      'centerItems',
      $style.buttonLink,
      {
        [$style.alignRight]: icon && iconPosition === 'right',
        [$style.disabled]: disabled,
        'centerAll fullWidth': fullWidth,
      },
    ]"
    :disabled
    draggable="false"
    :type
    v-bind="$attrs"
  >
    <component
      :is="icon"
      v-if="icon"
      ref="iconComponent"
      aria-hidden="true"
      :class="$style.icon"
      :color="iconColor"
      :size="ICON_SIZE[iconSize]"
      :weight="iconWeight"
    />

    <span
      ref="text"
      :class="[
        'clamp',
        $style.text,
        {
          visuallyHidden: !showText,
        },
      ]"
    >
      <slot />
    </span>
  </component>
</template>

<style module>
.buttonLink {
  position: relative;
  gap: var(--default-space);
  padding: var(--default-space);
  white-space: nowrap;
}

.alignRight {
  flex-direction: row-reverse;
}

.disabled {
  pointer-events: none;
  opacity: 0.25;
}

.icon {
  flex-shrink: 0;

  [opacity='0.2'] {
    opacity: 0.4;
    fill: var(--secondary-font-color);
  }
}

.text {
  display: block;
  margin-bottom: 1px;
}
</style>
