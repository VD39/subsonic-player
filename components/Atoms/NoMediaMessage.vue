<script setup lang="ts">
defineProps<{
  icon: string;
  message: string;
}>();
</script>

<template>
  <div :class="$style.noMediaMessage">
    <div :class="['column', 'centerItems', $style.content]">
      <component :is="icon" :size="50" />

      <p>{{ message }}</p>
    </div>
  </div>
</template>

<style module>
.noMediaMessage {
  position: relative;
  height: var(--width-height-100);
}

.content {
  position: absolute;
  inset: 50% auto auto 50%;
  transform: translate(-50%, -50%);
}
</style>
