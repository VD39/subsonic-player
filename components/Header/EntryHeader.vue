<script setup lang="ts">
import PreloadImage from '@/components/Image/PreloadImage.vue';

defineProps<{
  images: string[];
  title: string;
}>();
</script>

<template>
  <div :class="$style.background">
    <PreloadImage
      :image="images[0]"
      :alt="title"
      :class="$style.backgroundImage"
    />
    <div :class="$style.blur" />
  </div>

  <section :class="$style.content">
    <figure :class="$style.figure">
      <PreloadImage
        v-for="image in images"
        :key="image"
        :image="image"
        :class="$style.image"
        :alt="title"
        data-test-id="image"
      />
    </figure>

    <div>
      <h1>{{ title }}</h1>

      <slot />
    </div>
  </section>
</template>

<style module>
.background {
  position: absolute;
  inset: 0 0 auto;
  z-index: 0;
  height: 500px;
  overflow: hidden;
}

.backgroundImage {
  position: absolute;
  inset: -20% auto auto -20%;
  width: 120%;
  height: 120%;
  filter: brightness(0.75) blur(3.5rem);
}

.blur {
  position: absolute;
  inset: auto auto 0 0;
  width: 100%;
  height: 100%;
  background-image: var(--image-blur);
}

.content {
  position: relative;
  display: flex;
  gap: var(--space-24);
  margin-bottom: var(--space-40);
}

.figure {
  position: relative;
  display: grid;
  flex-shrink: 0;
  gap: var(--space-4);
  width: 300px;
  height: 300px;
  aspect-ratio: 1;
  padding: var(--space-4);
  overflow: hidden;
  background-image: var(--image-blur);

  :nth-child(2) {
    grid-column-start: 2;
  }

  :nth-child(1):nth-last-child(3) {
    grid-row-start: span 2;
  }
}

.image {
  width: 100%;
  height: 100%;
}
</style>
