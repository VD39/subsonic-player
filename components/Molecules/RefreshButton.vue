<script setup lang="ts">
import ButtonLink from '@/components/Atoms/ButtonLink.vue';
import SpinningLoader from '@/components/Atoms/SpinningLoader.vue';

defineEmits(['refresh']);

const loading = useLoading();

const buttonProps = computed<ButtonProps>(() => ({
  icon: loading.value ? SpinningLoader : ICONS.refresh,
  text: `${loading.value ? 'Refreshing' : 'Refresh'} page data`,
}));
</script>

<template>
  <ButtonLink
    :icon="buttonProps.icon"
    :title="buttonProps.text"
    icon-size="large"
    @click="$emit('refresh')"
  >
    {{ buttonProps.text }}
  </ButtonLink>
</template>
