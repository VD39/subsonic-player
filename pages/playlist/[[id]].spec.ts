import type { VueWrapper } from '@vue/test-utils';

import { mockNuxtImport } from '@nuxt/test-utils/runtime';
import { mount } from '@vue/test-utils';

import ButtonLink from '@/components/Atoms/ButtonLink.vue';
import NoMediaMessage from '@/components/Atoms/NoMediaMessage.vue';
import DropdownMenu from '@/components/Molecules/Dropdown/DropdownMenu.vue';
import RefreshButton from '@/components/Molecules/RefreshButton.vue';
import EntryHeader from '@/components/Organisms/EntryHeader.vue';
import MixedTracksList from '@/components/Organisms/TrackLists/MixedTracksList.vue';
import { getFormattedPlaylistsMock } from '@/test/helpers';
import { useAudioPlayerMock } from '@/test/useAudioPlayerMock';
import { useHeadMock } from '@/test/useHeadMock';

import PlaylistPage from './[[id]].vue';

const downloadMediaMock = vi.fn();

mockNuxtImport('useMediaLibrary', () => () => ({
  downloadMedia: downloadMediaMock,
}));

const openTrackInformationModalMock = vi.fn();

mockNuxtImport('useMediaInformation', () => () => ({
  openTrackInformationModal: openTrackInformationModalMock,
}));

const addToPlaylistModalMock = vi.fn();
const deletePlaylistMock = vi.fn();
const getPlaylistTracksByIdMock = vi.fn();
const removeFromPlaylistMock = vi.fn();
const updatePlaylistModalMock = vi.fn();

const playlistMock = ref<null | Playlist>(null);

mockNuxtImport('usePlaylist', () => () => ({
  addToPlaylistModal: addToPlaylistModalMock,
  deletePlaylist: deletePlaylistMock,
  getPlaylistTracksById: getPlaylistTracksByIdMock,
  playlist: playlistMock,
  removeFromPlaylist: removeFromPlaylistMock,
  updatePlaylistModal: updatePlaylistModalMock,
}));

const dragStartMock = vi.fn();

mockNuxtImport('useDragAndDrop', () => () => ({
  dragStart: dragStartMock,
}));

const refreshMock = vi.fn();

mockNuxtImport('useAsyncData', () => () => ({
  refresh: refreshMock,
  status: ref('success'),
}));

const { routeMock } = vi.hoisted(() => ({
  routeMock: vi.fn(() => ({
    fullPath: '/playlist/0',
    params: {
      id: 0,
    },
  })),
}));

mockNuxtImport('useRoute', () => routeMock);

const navigateToMock = vi.hoisted(() => vi.fn());

mockNuxtImport('navigateTo', () => navigateToMock);

const { useHeadTitleMock } = useHeadMock();
const {
  addTracksToQueueMock,
  addTrackToQueueMock,
  playTracksMock,
  shuffleTracksMock,
} = useAudioPlayerMock();

const track = getFormattedPlaylistsMock()[0].tracks[0];

async function factory(props = {}) {
  const wrapper = mount(PlaylistPage, {
    global: {
      stubs: {
        MixedTracksList: true,
      },
    },
    props: {
      ...props,
    },
  });

  await wrapper.vm.$nextTick();

  const dropdownMenu = wrapper.findComponent(DropdownMenu);

  if (dropdownMenu.exists()) {
    dropdownMenu.findComponent(ButtonLink).vm.$emit('click');
  }

  return wrapper;
}

describe('[[id]]', () => {
  let wrapper: VueWrapper;

  beforeEach(async () => {
    wrapper = await factory();
  });

  afterEach(() => {
    vi.clearAllMocks();
  });

  it('matches the snapshot', () => {
    expect(wrapper.html()).toMatchSnapshot();
  });

  describe('when getPlaylistTracksById does not return any data', () => {
    beforeEach(async () => {
      playlistMock.value = null;
      wrapper = await factory();
    });

    it('sets the useHead function with correct title', () => {
      expect(useHeadTitleMock.value).toBe('Playlist');
    });

    it('shows the NoMediaMessage component', () => {
      expect(wrapper.findComponent(NoMediaMessage).exists()).toBe(true);
    });

    it('does not show the playlist content', () => {
      expect(wrapper.find({ ref: 'playlistContent' }).exists()).toBe(false);
    });
  });

  describe('when getPlaylistTracksById does return data', () => {
    beforeEach(async () => {
      playlistMock.value = getFormattedPlaylistsMock()[0];
      wrapper = await factory();
    });

    it('matches the snapshot', () => {
      expect(wrapper.html()).toMatchSnapshot();
    });

    it('sets the useHead function with correct title', () => {
      expect(useHeadTitleMock.value).toBe('playlist-0 - Playlist');
    });

    it('shows the playlist content', () => {
      expect(wrapper.find({ ref: 'playlistContent' }).exists()).toBe(true);
    });

    it('does not show the NoMediaMessage component', () => {
      expect(wrapper.findComponent(NoMediaMessage).exists()).toBe(false);
    });

    describe('when playlist.trackCount is 1', () => {
      it('shows the correct track count text', () => {
        expect(wrapper.find({ ref: 'trackCount' }).text()).toBe('1 Track');
      });
    });

    describe('when playlist.trackCount is greater than 1', () => {
      beforeEach(async () => {
        playlistMock.value = getFormattedPlaylistsMock(1, {
          trackCount: 5,
        })[0];

        wrapper = await factory();
      });

      it('matches the snapshot', () => {
        expect(wrapper.html()).toMatchSnapshot();
      });

      it('shows the correct track count text', () => {
        expect(wrapper.find({ ref: 'trackCount' }).text()).toBe('5 Tracks');
      });
    });

    describe('when the update playlist DropdownItem component emits a click event', () => {
      beforeEach(() => {
        wrapper
          .findComponent({ ref: 'updatePlaylistDropdownItem' })
          .vm.$emit('click');
      });

      it('calls the updatePlaylistModal function with correct parameters', () => {
        expect(updatePlaylistModalMock).toHaveBeenCalledWith(
          playlistMock.value,
        );
      });
    });

    describe('when the delete playlist DropdownItem component emits a click event', () => {
      beforeEach(() => {
        wrapper
          .findComponent({ ref: 'deletePlaylistDropdownItem' })
          .vm.$emit('click');
      });

      it('calls the deletePlaylist function with correct parameters', () => {
        expect(deletePlaylistMock).toHaveBeenCalledWith(0);
      });

      it('calls the navigateTo function with correct parameters', () => {
        expect(navigateToMock).toHaveBeenCalledWith({
          name: ROUTE_NAMES.playlists,
        });
      });
    });

    describe('when playlist tracks is an empty array', () => {
      beforeEach(async () => {
        playlistMock.value = getFormattedPlaylistsMock(1, {
          tracks: [],
        })[0];

        wrapper = await factory();
      });

      it('matches the snapshot', () => {
        expect(wrapper.html()).toMatchSnapshot();
      });

      it('sets the correct disabled prop on the play tracks ButtonLink component', () => {
        expect(
          wrapper.findComponent({ ref: 'playTracksButton' }).props('disabled'),
        ).toBe(true);
      });

      it('sets the correct disabled prop on the shuffle track ButtonLink component', () => {
        expect(
          wrapper
            .findComponent({ ref: 'shuffleTracksButton' })
            .props('disabled'),
        ).toBe(true);
      });

      it('does not show the add to queue DropdownItem component', () => {
        expect(
          wrapper.findComponent({ ref: 'addToQueueDropdownItem' }).exists(),
        ).toBe(false);
      });

      it('does not show the play tracks DropdownItem component', () => {
        expect(
          wrapper.findComponent({ ref: 'playTracksDropdownItem' }).exists(),
        ).toBe(false);
      });
    });

    describe('when playlist tracks is not an empty array', () => {
      it('sets the correct disabled prop on the play tracks ButtonLink component', () => {
        expect(
          wrapper.findComponent({ ref: 'playTracksButton' }).props('disabled'),
        ).toBe(false);
      });

      it('sets the correct disabled prop on the shuffle track ButtonLink component', () => {
        expect(
          wrapper
            .findComponent({ ref: 'shuffleTracksButton' })
            .props('disabled'),
        ).toBe(false);
      });

      it('shows the add to queue DropdownItem component', () => {
        expect(
          wrapper.findComponent({ ref: 'addToQueueDropdownItem' }).exists(),
        ).toBe(true);
      });

      it('shows the play tracks DropdownItem component', () => {
        expect(
          wrapper.findComponent({ ref: 'playTracksDropdownItem' }).exists(),
        ).toBe(true);
      });

      describe('when the play tracks ButtonLink component emits a click event', () => {
        beforeEach(() => {
          wrapper.findComponent({ ref: 'playTracksButton' }).vm.$emit('click');
        });

        it('calls the playTracks function with correct parameters', () => {
          expect(playTracksMock).toHaveBeenCalledWith(
            playlistMock.value!.tracks,
          );
        });
      });

      describe('when the shuffle tracks ButtonLink component emits a click event', () => {
        beforeEach(() => {
          wrapper
            .findComponent({ ref: 'shuffleTracksButton' })
            .vm.$emit('click');
        });

        it('calls the shuffleTracks function with correct parameters', () => {
          expect(shuffleTracksMock).toHaveBeenCalledWith(
            playlistMock.value!.tracks,
          );
        });
      });

      describe('when the add to queue DropdownItem component emits a click event', () => {
        beforeEach(() => {
          wrapper
            .findComponent({ ref: 'addToQueueDropdownItem' })
            .vm.$emit('click');
        });

        it('calls the addTracksToQueue function with correct parameters', () => {
          expect(addTracksToQueueMock).toHaveBeenCalledWith(
            playlistMock.value!.tracks,
          );
        });
      });

      describe('when the play tracks DropdownItem component emits a click event', () => {
        beforeEach(() => {
          wrapper
            .findComponent({ ref: 'playTracksDropdownItem' })
            .vm.$emit('click');
        });

        it('calls the playTracks function with correct parameters', () => {
          expect(playTracksMock).toHaveBeenCalledWith(
            playlistMock.value!.tracks,
          );
        });
      });
    });

    describe(`when playlist.id is not equal to ${RANDOM_PLAYLIST.id}`, () => {
      it('sets the correct hideRemoveOption prop on the MixedTracksList component', () => {
        expect(
          wrapper.findComponent(MixedTracksList).props('hideRemoveOption'),
        ).toBe(false);
      });
    });

    describe(`when playlist.id is equal to ${RANDOM_PLAYLIST.id}`, () => {
      beforeEach(async () => {
        playlistMock.value = {
          ...getFormattedPlaylistsMock(1, {})[0],
          id: RANDOM_PLAYLIST.id,
        };

        wrapper = await factory();
      });

      it('sets the correct hideRemoveOption prop on the MixedTracksList component', () => {
        expect(
          wrapper.findComponent(MixedTracksList).props('hideRemoveOption'),
        ).toBe(true);
      });
    });

    describe('when the EntryHeader component emits the dragStart event', () => {
      beforeEach(() => {
        wrapper.findComponent(EntryHeader).vm.$emit('dragStart', DragEvent);
      });

      it('calls the dragStart function with correct parameters', () => {
        expect(dragStartMock).toHaveBeenCalledWith(
          playlistMock.value,
          DragEvent,
        );
      });
    });

    describe('when the RefreshButton component emits the refresh event', () => {
      beforeEach(() => {
        wrapper.findComponent(RefreshButton).vm.$emit('refresh');
      });

      it('calls the refresh function', () => {
        expect(refreshMock).toHaveBeenCalled();
      });
    });

    describe('when the MixedTracksList component emits the addToPlaylist event', () => {
      beforeEach(() => {
        wrapper.findComponent(MixedTracksList).vm.$emit('addToPlaylist', track);
      });

      it('calls the addToPlaylistModal function with the correct parameters', () => {
        expect(addToPlaylistModalMock).toHaveBeenCalledWith(track);
      });
    });

    describe('when the MixedTracksList component emits the addToQueue event', () => {
      beforeEach(() => {
        wrapper.findComponent(MixedTracksList).vm.$emit('addToQueue', track);
      });

      it('calls the addTrackToQueue function with the correct parameters', () => {
        expect(addTrackToQueueMock).toHaveBeenCalledWith(track);
      });
    });

    describe('when the MixedTracksList component emits the downloadMedia event', () => {
      beforeEach(() => {
        wrapper.findComponent(MixedTracksList).vm.$emit('downloadMedia', track);
      });

      it('calls the downloadMedia function with the correct parameters', () => {
        expect(downloadMediaMock).toHaveBeenCalledWith(track);
      });
    });

    describe('when the MixedTracksList component emits the dragStart event', () => {
      beforeEach(() => {
        wrapper.findComponent(MixedTracksList).vm.$emit('dragStart', track);
      });

      it('calls the dragStart function with the correct parameters', () => {
        expect(dragStartMock).toHaveBeenCalledWith(track);
      });
    });

    describe('when the MixedTracksList component emits the mediaInformation event', () => {
      beforeEach(() => {
        wrapper
          .findComponent(MixedTracksList)
          .vm.$emit('mediaInformation', track);
      });

      it('calls the openTrackInformationModal function with the correct parameters', () => {
        expect(openTrackInformationModalMock).toHaveBeenCalledWith(track);
      });
    });

    describe('when the MixedTracksList component emits the playTrack event', () => {
      beforeEach(() => {
        wrapper.findComponent(MixedTracksList).vm.$emit('playTrack', 1);
      });

      it('calls the playTracks function with the correct parameters', () => {
        expect(playTracksMock).toHaveBeenCalledWith(
          playlistMock.value!.tracks,
          0,
        );
      });
    });

    describe('when the MixedTracksList component emits the remove event', () => {
      beforeEach(() => {
        wrapper.findComponent(MixedTracksList).vm.$emit('remove', {
          index: 1,
        });
      });

      it('calls the removeFromPlaylist function with correct parameters', () => {
        expect(removeFromPlaylistMock).toHaveBeenCalledWith({
          playlistId: 0,
          songIndexToRemove: 1,
        });
      });
    });
  });
});
