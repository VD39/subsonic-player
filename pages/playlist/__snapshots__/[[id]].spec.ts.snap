// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`[[id]] > matches the snapshot 1`] = `
"<div class="column centerAll loadingData" style="display: none;"><span class="mainLoader"><span class="headphones"><svg fill="none" stroke="currentColor" stroke-linecap="round" stroke-linejoin="round" stroke-width="1.25" viewBox="0 0 24 24" width="75" xmlns="http://www.w3.org/2000/svg"><path d="M4 13.5V13c0-4.97 3.582-9 8-9s8 4.03 8 9v.5"></path><path d="M2 17.439v-1.877a2 2 0 0 1 1.515-1.94L4 13.5l1.254-.314a.6.6 0 0 1 .746.582v5.464a.6.6 0 0 1-.746.582l-1.74-.435A2 2 0 0 1 2 17.439ZM22 17.439v-1.877a2 2 0 0 0-1.515-1.94L20 13.5l-1.255-.314a.6.6 0 0 0-.745.582v5.464a.6.6 0 0 0 .745.582l1.74-.435A2 2 0 0 0 22 17.439Z"></path></svg></span><span class="centerItems loader"><span class="bar"></span><span class="bar"></span><span class="bar"></span><span class="bar"></span></span><span class="visuallyHidden">Loading data</span></span></div>
<div class="mainContent">
  <div class="column centerAll noMediaMessage">
    <phplaylist size="50"></phplaylist>
    <p>No playlist found.</p>
  </div>
</div>"
`;

exports[`[[id]] > when getPlaylistTracksById does return data > matches the snapshot 1`] = `
"<div class="column centerAll loadingData" style="display: none;"><span class="mainLoader"><span class="headphones"><svg fill="none" stroke="currentColor" stroke-linecap="round" stroke-linejoin="round" stroke-width="1.25" viewBox="0 0 24 24" width="75" xmlns="http://www.w3.org/2000/svg"><path d="M4 13.5V13c0-4.97 3.582-9 8-9s8 4.03 8 9v.5"></path><path d="M2 17.439v-1.877a2 2 0 0 1 1.515-1.94L4 13.5l1.254-.314a.6.6 0 0 1 .746.582v5.464a.6.6 0 0 1-.746.582l-1.74-.435A2 2 0 0 1 2 17.439ZM22 17.439v-1.877a2 2 0 0 0-1.515-1.94L20 13.5l-1.255-.314a.6.6 0 0 0-.745.582v5.464a.6.6 0 0 0 .745.582l1.74-.435A2 2 0 0 0 22 17.439Z"></path></svg></span><span class="centerItems loader"><span class="bar"></span><span class="bar"></span><span class="bar"></span><span class="bar"></span></span><span class="visuallyHidden">Loading data</span></span></div>
<div class="mainContent">
  <div>
    <div class="entryHeader">
      <div class="centerAll preloadImage backgroundImage"><span class="skeletonLoader"><span class="visuallyHidden">Loading image</span></span><img alt="playlist-0" class="image" draggable="false" loading="lazy" src="image"></div>
      <div class="blur"></div>
    </div>
    <section class="mBXL content">
      <figure class="figure" draggable="true">
        <div class="centerAll preloadImage preloadImage" data-test-id="image"><span class="skeletonLoader"><span class="visuallyHidden">Loading image</span></span><img alt="playlist-0" class="image" draggable="false" loading="lazy" src="image"></div>
      </figure>
      <div class="mBAllM detailsWrapper">
        <div class="centerItems spaceBetween">
          <h1>playlist-0</h1>
          <div class="centerItems"><button class="centerItems buttonLink" draggable="false" type="button" id="refresh-data-button" title="Refresh page data">
              <pharrowsclockwise aria-hidden="true" class="icon" size="32" weight="regular"></pharrowsclockwise><span class="clamp text visuallyHidden">Refresh page data</span>
            </button></div>
        </div>
        <ul class="bulletList">
          <li>Playlist</li>
          <li><span class="strong">1</span> Track</li>
          <li><span class="visuallyHidden">Duration: </span><time>00:01</time></li>
        </ul>
        <div class="list"><button class="centerItems buttonLink largeThemeHoverButton largeThemeHoverButton" draggable="false" type="button" id="play-all-button" title="Play tracks">
            <phplay aria-hidden="true" class="icon" size="24" weight="regular"></phplay><span class="clamp text visuallyHidden"> Play tracks </span>
          </button><button class="centerItems buttonLink" draggable="false" type="button" id="shuffle-all-button" title="Shuffle tracks">
            <phshuffle aria-hidden="true" class="icon" size="24" weight="regular"></phshuffle><span class="clamp text visuallyHidden"> Shuffle tracks </span>
          </button>
          <div class="centerItems dropdownMenu"><button class="centerItems buttonLink alignRight buttonLink buttonLink" draggable="false" type="button" title="More">
              <phdotsthreevertical aria-hidden="true" class="icon" size="24" weight="regular"></phdotsthreevertical><span class="clamp text visuallyHidden">More</span>
            </button>
            <div class="dropdown fade-enter-from fade-enter-active">
              <ul>
                <li class="dropdownItem"><button class="centerItems buttonLink centerAll fullWidth spaceBetween buttonLink spaceBetween buttonLink" draggable="false" type="button">
                    <!--v-if--><span class="clamp text"> Edit Playlist </span>
                  </button></li>
                <li class="dropdownItem"><button class="centerItems buttonLink centerAll fullWidth spaceBetween buttonLink spaceBetween buttonLink" draggable="false" type="button">
                    <!--v-if--><span class="clamp text"> Delete Playlist </span>
                  </button></li>
                <li aria-hidden="true" class="divider"></li>
                <li class="dropdownItem"><button class="centerItems buttonLink centerAll fullWidth spaceBetween buttonLink spaceBetween buttonLink" draggable="false" type="button">
                    <!--v-if--><span class="clamp text"> Add to queue </span>
                  </button></li>
                <li class="dropdownItem"><button class="centerItems buttonLink centerAll fullWidth spaceBetween buttonLink spaceBetween buttonLink" draggable="false" type="button">
                    <!--v-if--><span class="clamp text"> Play Tracks </span>
                  </button></li>
              </ul>
            </div>
          </div>
        </div>
      </div>
    </section>
    <mixed-tracks-list-stub tracks="[object Object]" hideremoveoption="false"></mixed-tracks-list-stub>
  </div>
</div>"
`;

exports[`[[id]] > when getPlaylistTracksById does return data > when playlist tracks is an empty array > matches the snapshot 1`] = `
"<div class="column centerAll loadingData" style="display: none;"><span class="mainLoader"><span class="headphones"><svg fill="none" stroke="currentColor" stroke-linecap="round" stroke-linejoin="round" stroke-width="1.25" viewBox="0 0 24 24" width="75" xmlns="http://www.w3.org/2000/svg"><path d="M4 13.5V13c0-4.97 3.582-9 8-9s8 4.03 8 9v.5"></path><path d="M2 17.439v-1.877a2 2 0 0 1 1.515-1.94L4 13.5l1.254-.314a.6.6 0 0 1 .746.582v5.464a.6.6 0 0 1-.746.582l-1.74-.435A2 2 0 0 1 2 17.439ZM22 17.439v-1.877a2 2 0 0 0-1.515-1.94L20 13.5l-1.255-.314a.6.6 0 0 0-.745.582v5.464a.6.6 0 0 0 .745.582l1.74-.435A2 2 0 0 0 22 17.439Z"></path></svg></span><span class="centerItems loader"><span class="bar"></span><span class="bar"></span><span class="bar"></span><span class="bar"></span></span><span class="visuallyHidden">Loading data</span></span></div>
<div class="mainContent">
  <div>
    <div class="entryHeader">
      <div class="centerAll preloadImage backgroundImage"><span class="skeletonLoader"><span class="visuallyHidden">Loading image</span></span><img alt="playlist-0" class="image" draggable="false" loading="lazy" src="image"></div>
      <div class="blur"></div>
    </div>
    <section class="mBXL content">
      <figure class="figure" draggable="true">
        <div class="centerAll preloadImage preloadImage" data-test-id="image"><span class="skeletonLoader"><span class="visuallyHidden">Loading image</span></span><img alt="playlist-0" class="image" draggable="false" loading="lazy" src="image"></div>
      </figure>
      <div class="mBAllM detailsWrapper">
        <div class="centerItems spaceBetween">
          <h1>playlist-0</h1>
          <div class="centerItems"><button class="centerItems buttonLink" draggable="false" type="button" id="refresh-data-button" title="Refresh page data">
              <pharrowsclockwise aria-hidden="true" class="icon" size="32" weight="regular"></pharrowsclockwise><span class="clamp text visuallyHidden">Refresh page data</span>
            </button></div>
        </div>
        <ul class="bulletList">
          <li>Playlist</li>
          <li><span class="strong">1</span> Track</li>
          <li><span class="visuallyHidden">Duration: </span><time>00:01</time></li>
        </ul>
        <div class="list"><button class="centerItems buttonLink disabled largeThemeHoverButton largeThemeHoverButton" disabled="" draggable="false" type="button" id="play-all-button" title="Play tracks">
            <phplay aria-hidden="true" class="icon" size="24" weight="regular"></phplay><span class="clamp text visuallyHidden"> Play tracks </span>
          </button><button class="centerItems buttonLink disabled" disabled="" draggable="false" type="button" id="shuffle-all-button" title="Shuffle tracks">
            <phshuffle aria-hidden="true" class="icon" size="24" weight="regular"></phshuffle><span class="clamp text visuallyHidden"> Shuffle tracks </span>
          </button>
          <div class="centerItems dropdownMenu"><button class="centerItems buttonLink alignRight buttonLink buttonLink" draggable="false" type="button" title="More">
              <phdotsthreevertical aria-hidden="true" class="icon" size="24" weight="regular"></phdotsthreevertical><span class="clamp text visuallyHidden">More</span>
            </button>
            <div class="dropdown fade-enter-from fade-enter-active">
              <ul>
                <li class="dropdownItem"><button class="centerItems buttonLink centerAll fullWidth spaceBetween buttonLink spaceBetween buttonLink" draggable="false" type="button">
                    <!--v-if--><span class="clamp text"> Edit Playlist </span>
                  </button></li>
                <li class="dropdownItem"><button class="centerItems buttonLink centerAll fullWidth spaceBetween buttonLink spaceBetween buttonLink" draggable="false" type="button">
                    <!--v-if--><span class="clamp text"> Delete Playlist </span>
                  </button></li>
                <!--v-if-->
              </ul>
            </div>
          </div>
        </div>
      </div>
    </section>
    <mixed-tracks-list-stub tracks="" hideremoveoption="false"></mixed-tracks-list-stub>
  </div>
</div>"
`;

exports[`[[id]] > when getPlaylistTracksById does return data > when playlist.trackCount is greater than 1 > matches the snapshot 1`] = `
"<div class="column centerAll loadingData" style="display: none;"><span class="mainLoader"><span class="headphones"><svg fill="none" stroke="currentColor" stroke-linecap="round" stroke-linejoin="round" stroke-width="1.25" viewBox="0 0 24 24" width="75" xmlns="http://www.w3.org/2000/svg"><path d="M4 13.5V13c0-4.97 3.582-9 8-9s8 4.03 8 9v.5"></path><path d="M2 17.439v-1.877a2 2 0 0 1 1.515-1.94L4 13.5l1.254-.314a.6.6 0 0 1 .746.582v5.464a.6.6 0 0 1-.746.582l-1.74-.435A2 2 0 0 1 2 17.439ZM22 17.439v-1.877a2 2 0 0 0-1.515-1.94L20 13.5l-1.255-.314a.6.6 0 0 0-.745.582v5.464a.6.6 0 0 0 .745.582l1.74-.435A2 2 0 0 0 22 17.439Z"></path></svg></span><span class="centerItems loader"><span class="bar"></span><span class="bar"></span><span class="bar"></span><span class="bar"></span></span><span class="visuallyHidden">Loading data</span></span></div>
<div class="mainContent">
  <div>
    <div class="entryHeader">
      <div class="centerAll preloadImage backgroundImage"><span class="skeletonLoader"><span class="visuallyHidden">Loading image</span></span><img alt="playlist-0" class="image" draggable="false" loading="lazy" src="image"></div>
      <div class="blur"></div>
    </div>
    <section class="mBXL content">
      <figure class="figure" draggable="true">
        <div class="centerAll preloadImage preloadImage" data-test-id="image"><span class="skeletonLoader"><span class="visuallyHidden">Loading image</span></span><img alt="playlist-0" class="image" draggable="false" loading="lazy" src="image"></div>
      </figure>
      <div class="mBAllM detailsWrapper">
        <div class="centerItems spaceBetween">
          <h1>playlist-0</h1>
          <div class="centerItems"><button class="centerItems buttonLink" draggable="false" type="button" id="refresh-data-button" title="Refresh page data">
              <pharrowsclockwise aria-hidden="true" class="icon" size="32" weight="regular"></pharrowsclockwise><span class="clamp text visuallyHidden">Refresh page data</span>
            </button></div>
        </div>
        <ul class="bulletList">
          <li>Playlist</li>
          <li><span class="strong">5</span> Tracks</li>
          <li><span class="visuallyHidden">Duration: </span><time>00:01</time></li>
        </ul>
        <div class="list"><button class="centerItems buttonLink largeThemeHoverButton largeThemeHoverButton" draggable="false" type="button" id="play-all-button" title="Play tracks">
            <phplay aria-hidden="true" class="icon" size="24" weight="regular"></phplay><span class="clamp text visuallyHidden"> Play tracks </span>
          </button><button class="centerItems buttonLink" draggable="false" type="button" id="shuffle-all-button" title="Shuffle tracks">
            <phshuffle aria-hidden="true" class="icon" size="24" weight="regular"></phshuffle><span class="clamp text visuallyHidden"> Shuffle tracks </span>
          </button>
          <div class="centerItems dropdownMenu"><button class="centerItems buttonLink alignRight buttonLink buttonLink" draggable="false" type="button" title="More">
              <phdotsthreevertical aria-hidden="true" class="icon" size="24" weight="regular"></phdotsthreevertical><span class="clamp text visuallyHidden">More</span>
            </button>
            <div class="dropdown fade-enter-from fade-enter-active">
              <ul>
                <li class="dropdownItem"><button class="centerItems buttonLink centerAll fullWidth spaceBetween buttonLink spaceBetween buttonLink" draggable="false" type="button">
                    <!--v-if--><span class="clamp text"> Edit Playlist </span>
                  </button></li>
                <li class="dropdownItem"><button class="centerItems buttonLink centerAll fullWidth spaceBetween buttonLink spaceBetween buttonLink" draggable="false" type="button">
                    <!--v-if--><span class="clamp text"> Delete Playlist </span>
                  </button></li>
                <li aria-hidden="true" class="divider"></li>
                <li class="dropdownItem"><button class="centerItems buttonLink centerAll fullWidth spaceBetween buttonLink spaceBetween buttonLink" draggable="false" type="button">
                    <!--v-if--><span class="clamp text"> Add to queue </span>
                  </button></li>
                <li class="dropdownItem"><button class="centerItems buttonLink centerAll fullWidth spaceBetween buttonLink spaceBetween buttonLink" draggable="false" type="button">
                    <!--v-if--><span class="clamp text"> Play Tracks </span>
                  </button></li>
              </ul>
            </div>
          </div>
        </div>
      </div>
    </section>
    <mixed-tracks-list-stub tracks="[object Object]" hideremoveoption="false"></mixed-tracks-list-stub>
  </div>
</div>"
`;
